pupil_src/shared_modules/calibration_routines/calibrate_2d.py

```diff
@@ -0,0 +1,264 @@
+        n_params,
+        calibration_data[:, :-2],
+        calibration_data[:, -2],
+        binocular,
+        dof=dof,
+        degree=degree,
+        ignored_terms=ignored_terms,
+        regularization=regularization,
+        loss_scale=loss_scale,
+    )
+
+    success_y, params_y = optimize_parameters(
+        n_params,
+        calibration_data[:, :-2],
+        calibration_data[:, -1],
+        binocular,
+        dof=dof,
+        degree=degree,
+        ignored_terms=ignored_terms,
+        regularization=regularization,
+        loss_scale=loss_scale,
+    )
+
+    return (success_x and success_y), (params_x, params_y, dof, degree, binocular)
+
+
+def optimize_parameters(
+    n_params,
+    pupil_positions,
+    targets,
+    binocular,
+    dof=2,
+    degree=2,
+    ignored_terms=(),
+    regularization=0.0,
+    loss_scale=-1,
+):
+
+    initial_guess = np.zeros(n_params)
+
+    if (0, 0) not in ignored_terms:
+        initial_guess[0] = np.mean(targets)
+    if binocular:
+        initial_guess[n_params // 2] = initial_guess[0]
+
+    fit_result = scipy.optimize.minimize(
+        calibration_cost,
+        initial_guess,
+        args=(
+            pupil_positions,
+            targets,
+            binocular,
+            dof,
+            degree,
+            ignored_terms,
+            regularization,
+            loss_scale,
+        ),
+    )
+
+    final_params = list(
+        extend_params(fit_result.x, degree, ignored_terms, 0, binocular)
+    )
+
+    return fit_result.success, final_params
+
+
+def calibration_cost(
+    params,
+    pupil_positions,
+    targets,
+    binocular,
+    dof=2,
+    degree=2,
+    ignored_terms=(),
+    regularization=0.0,
+    loss_scale=-1,
+):
+    extended_params = extend_params(params, degree, ignored_terms, 0, binocular)
+
+    if binocular:
+        polynomial_ = PolynomialBinocular(extended_params, dof, degree)
+    else:
+        polynomial_ = PolynomialMonocular(extended_params, dof, degree)
+
+    squared_residuals = (
+        polynomial_([column for column in pupil_positions.T]) - targets
+    ) ** 2
+
+    if loss_scale > 0:
+        cost = 1 / 2 * np.sum(cauchy_loss(squared_residuals, loss_scale))
+    else:
+        cost = 1 / 2 * np.sum(squared_residuals)
+
+    cost += regularization * np.sum(params ** 2)
+
+    return cost
+
+
+def cauchy_loss(residuals, loss_scale):
+    return loss_scale ** 2 * np.log(1 + residuals / loss_scale ** 2)
+
+
+def convert_to_linear_index(multi_idx, size_per_dim):
+    """
+    Converts the multi-index of a multi-dimensional array to the corresponding linear index of the unraveled array.
+    Assumes the array to have the same size in all dimensions.
+    """
+    dims = len(multi_idx)
+    linear_idx = 0
+    for k, idx in enumerate(multi_idx):
+        linear_idx += idx * (size_per_dim ** (dims - 1 - k))
+    return linear_idx
+
+
+def extend_params(params, degree, ignored_terms=(), fill_value=0, binocular=False):
```

Comment:
I feel like you should rename this function to something that make the first sentence of the docstring unnecessary.